# Exercise 2 - Keras Model

This Jupyter notebook works through defining a neural network model using the *Keras* machine learning framework, with an architecture similar to that described in the lesson content.  In this exercise we will not train the model, rather setup the architecture of the model to reflect the following image.

![neuralnet.jpg](neuralnet.jpg)

Import the required dependencies. We will require the *Sequential* class from **keras.models** and *Dense* from **keras.layers**.

In [ ]:
#!conda install tensorflow keras
#!pip install tensorflow keras

In [1]:
from keras.models import Sequential
from keras.layers import Dense

Using TensorFlow backend.


We will construct the model by passing a list of the layers as an argument to the Sequential model class.

The first layer of any neural network model is the input layer which simply receives information from the data source.  In this lesson we will be using the CIFAR10 dataset, where each input is one of 60 000 samples of a 32 x 32 pixel RGB image, of one of 10 different image categories or classes:

1. airplane
2. automobile
3. bird
4. cat
5. deer
6. dog
7. frog
8. horse
9. ship
10. truck

Thus the number of inputs $m=32\times32=1024$.  In the first layer we will select $p=500$ nodes and will use a ReLU activation function.  

The output layer contains the same number of nodes as classes (10) and will use a softmax activation function to scale the values to be between 0 and 1 and provide a probability distribution for the outputs.

In [2]:
model = Sequential([
    Dense(500, input_shape=(1024,), activation='relu'),
    Dense(10, activation='softmax')
    
])

Instructions for updating:
Colocations handled automatically by placer.


Now that we have defined the model we can confirm its structure, using the *summary* method:

In [3]:
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_1 (Dense)              (None, 500)               512500    
_________________________________________________________________
dense_2 (Dense)              (None, 10)                5010      
Total params: 517,510
Trainable params: 517,510
Non-trainable params: 0
_________________________________________________________________


Here we can see the neural network, with two layers and 517 510 trainable parameters or weights ($\Theta$).